docs: enhance package structure

hjhsalo opened the pull request:
* add initial Sphinx documentation structure, with license,
  authors, etc. (closes #1)
* add initial Travis CI configuration and tests
* add copyright header messages to each source code file

Signed-off-by: Harri Hirvonsalo <harri.hirvonsalo@cern.ch>

diegodelemos commented on run-tests.sh:
```diff
@@ -0,0 +1,30 @@
+check-manifest --ignore ".travis-*" && \
+sphinx-build -qnNW docs docs/_build/html && \
+python setup.py test && \
+sphinx-build -qnNW -b doctest docs docs/_build/doctest
```
Shall we follow the same approach as for `Invenio` for `doctest` execution? https://github.com/inveniosoftware/invenio-rest/issues/65, https://github.com/inveniosoftware/troubleshooting/issues/3

diegodelemos commented on MANIFEST.in:
```diff
@@ -0,0 +1,30 @@
+# granted to it by virtue of its status as an Intergovernmental Organization or
+# submit itself to any jurisdiction.
+
+include Dockerfile
```
`Dockerfile` file missing?

hjhsalo commented:
**What has been updated (21.4.2017)**
* I changed the structure of the module a bit and added a Alpine Linux based Dockerfile (python:2.7-alpine)
* Python code now greets and stores (appends) the greeting to a filepath given as function parameters.
  (Defaults to "Hello World!" and 'output/output.txt')
* Behaviour can be changed using environment variables (REANA_DEMO_NAME, REANA_DEMO_OUTPUTFILE).
  Checking for environment variables is done in Python code and not e.g. in Dockerfile.

**To test with Docker**
* Build and run with altering the greeting and outputfolder.
  _Note: If `output`-directory doesn't exist, making a local volume mount with `-v` option causes the `output`-folder to be created with root-permissions and unprivileged user inside Docker Container cannot write to it._
  ```
  $ docker build -t reanahub/reana-demo-helloworld .
  $ mkdir output
  $ docker run --rm -v $PWD/output/:/code/output \
  -e REANA_DEMO_NAME='REANA' \
  -e REANA_DEMO_OUTPUTFILE='output/output-alt.txt' \
  reanahub/reana-demo-helloworld
  ```
* Check the content of 'output/output-alt.txt' for the greeting.
  (Should contain "Hello, REANA!")
  ```
  $ cat output/output-alt.txt
  ```

---

**Notes and possible improvements**
* Naming of things could be changed a bit, e.g. `outputfile` -> `outputpath`. Any comments / preferences / conventions on this?
* There are three simple test cases, which are infact partly overlapping.
  Should I refactor them to one test case that covers both `hello()` and `main()`?
* Is it OK to add `.coveragerc` -file to exclude unnecessary checks such as `'if __name__ == "__main__":'` to reach 100% coverage?
* Had to use `os.path.dirname(__file__)` in `docs/conf.py` to get rid of IOErrors about not existing version.py -file.
  I don't know how portable this is between operating systems, but then again, we are not supporting Windows.
  This should work, but there are many (better?) ways to achieve the same end result.
  Should I change this to something else?
* I didn't specify a config.py as amount of configurable things is (still?) low.
  On the other hand, this demo-helloworld could serve as a reference structure for other reana-demo-examples and it might be beneficial to do everything by the book / de facto standards.
  The same reasoning applies for tests. Should I do this in some other way?

**Notes on Docker Image sizes**
I opted for Alpine Linux based image to shrink down the filesize of the resulting Docker Image.
To further slim down the image I only install reana-demo-helloworld module and not it's dependencies needed to run tests or build documentation.
Here is a summary for comparison of Debian based image vs Alpine based one and the impact of not installing by .[all]:
* Python:2.7 = 745MB
* Python:2.7-alpine = 144MB
* Python:2.7-alpine + pip install without [all] -flag = 78.6MB

Further slimming down the size of image might be possible. I have to look into it later.

hjhsalo commented:
BTW, should I include the instruction part in documentation?
I'm not yet sure how this will be integrated to actual reanahub/reana installation and left out any usage instructions from the documentation for this reason.

diegodelemos commented on reana_demo_helloworld/__main__.py:
```diff
@@ -0,0 +1,59 @@
+
+
+def main(args=None):
+    """The main routine."""
```
Note
```console
reana_demo_helloworld/__main__.py:31 in public function `main`:
        D401: First line should be in imperative mood; try rephrasing (found 'The')
```
`Launch the main routine.` or so 👍

diegodelemos commented on reana_demo_helloworld/__init__.py:
```diff
@@ -0,0 +1,33 @@
+
+from .version import __version__
+
+__all__ = ('__version__',
```
why not all in the same line? I don't think we are going to have more in the future here 😃

diegodelemos commented on Dockerfile:
```diff
@@ -0,0 +1,36 @@
+    chown -R reanauser:reanauser /code
+USER reanauser
+
+ENV REANA_DEMO_NAME="World"
```
Based on the discussion in `reana` chat I think we should stick to an example without using env variables, something more in the line of [yadage hello world example](http://yadage.readthedocs.io/en/latest/#hello-world-example) and leave env vars for the future.
Either we provide the entire command to the image as `python -m reana_demo_helloworld -p name -o output/output.txt` or we specify `python -m reana_demo_helloworld` as a `docker` `entrypoint` and then just provide `-p` and `-o` along with its values.

diegodelemos commented on Dockerfile:
```diff
@@ -0,0 +1,35 @@
+
+CMD [ "python", "reana_demo_helloworld/helloworld.py", \
+      "--name", "J. Doe", \
+      "--outputfile", "output/output.txt" ]
```
Some delay parameter would be nice! 🐢 💨 💨

diegodelemos commented on Dockerfile:
```diff
@@ -0,0 +1,35 @@
+
+RUN adduser -u 1000 -D -g '' reanauser && \
+    chown -R reanauser:reanauser /code
+USER reanauser
```
Since we are writing as root in `reana-workflow-engine-yadage` we need to run this container as root as well, just remove the `adduser` and `USER reanauser` we will fix it soon 🙃  thanks!
